[kubeadm] Add etcd L7 check on upgrade

detiber opened the pull request:
- Adds L7 check for kubeadm etcd static pod upgrade

Currently this breaks some of the tests, but I wanted to get some eyes on this before heading to bed for the night. I'm not sure of a way that we can keep both the separate etcd client configs and also generate the client configs from the manifests/certificates dir.

detiber committed:
[kubeadm] Add etcd L7 check on upgrade

- Adds L7 check for kubeadm etcd static pod upgrade

detiber commented:
I'm still having difficulty finding a good way to mock out the etcd client in the tests, currently I've reverted staticpods_test to something closer to what it was previously where it skipped the etcd upgrade bits

detiber commented:
@stealthybox This is currently at a point where it is passing my local upgrade testing and unit tests are passing. ptal

stealthybox commented on cmd/kubeadm/app/phases/upgrade/staticpods.go:
```diff
@@ -299,51 +301,95 @@ func performEtcdStaticPodUpgrade(waiter apiclient.Waiter, pathMgr StaticPodPathM
-				// Nothing else left to try to recover etcd cluster
-				return true, fmt.Errorf("fatal error upgrading local etcd cluster: %v, the backup of etcd database is stored here:(%s)", err, backupEtcdDir)
-			}
+	fmt.Println("[upgrade/etcd] waiting for etcd pod to become available after TLS upgrade")
```
Looks like this currently a general log statement for the L7 etcd check unrelated to pod status:

`s/etcd pod/etcd/`

`s/ after TLS upgrade//`
or
`if isTLSUpgrade {...}`

stealthybox commented on cmd/kubeadm/app/phases/upgrade/staticpods.go:
```diff
@@ -299,51 +301,95 @@ func performEtcdStaticPodUpgrade(waiter apiclient.Waiter, pathMgr StaticPodPathM
+	}
+
+	// Checking health state of etcd after the upgrade
+	delay := 30 * time.Second
```
Is this delay period still appropriate when the etcd image is pre-pulled?
Do you know how short an etcd startup could possibly be in an ideal case?

detiber replied:
The delay period is there to prevent the test from passing before the old etcd pod has been stopped by the kubelet. It could potentially be shortened, but that would depend on the kubelet polling interval for static pods and depend on system configuration/speed.

stealthybox replied:
Oh, interesting!
This is worth a comment mentioning that not sleeping could race for that reason.

What is the kubelet polling interval?
Should this be a constant or can we retrieve it from somewhere?

detiber replied:
I added a comment for this, I haven't dug into the kubelet polling interval. I tried to keep this simple for the bandaid backport, but I think using the kubelet api would remove the need for the delay for a more permanent fix.

stealthybox replied:
I think having a 30s initial delay regardless of whether `isTLSUpgrade` will probably be fine.

I'd be just as fine initializing it to something low like 2s and increasing it to 30s `if isTLSUpgrade`

stealthybox commented on cmd/kubeadm/app/phases/upgrade/staticpods.go:
```diff
@@ -299,51 +301,95 @@ func performEtcdStaticPodUpgrade(waiter apiclient.Waiter, pathMgr StaticPodPathM
+		if err != nil {
+			return true, fmt.Errorf("fatal error creating etcd client: %v", err)
+		}
+		newEtcdClient = client
```
honest question:
is there a good reason why this is not just
`newEtcdClient, err := etcdutil.NewStaticPodClient(` ?
is this for legibility?

detiber replied:
I tried this at first, and during testing locally I was seeing this create a new `newEtcdClient` variable in the conditional block scope and the `newEtcdClient` in the outer scope remained `nil`. I could have used `=` and pre-defined err, but this approach felt a bit cleaner to me.

stealthybox replied:
just learned something from this.
I was perplexed the other day about what the behavior for `:=` would be with a mixture of declared and undeclared vars.

detiber replied:
I'm still relatively new to Go, so this definitely caught me by surprise. It looks like := works just fine as long as any pre-defined vars are within the same block scope.

stealthybox commented on cmd/kubeadm/app/phases/upgrade/staticpods.go:
```diff
@@ -299,51 +301,95 @@ func performEtcdStaticPodUpgrade(waiter apiclient.Waiter, pathMgr StaticPodPathM
+	}
+
+	// etcd upgrade is done prior to other control plane components
+	if !isExternalEtcd && etcdUpgrade {
```
:+1: nice

stealthybox commented on cmd/kubeadm/app/phases/upgrade/staticpods.go:
```diff
@@ -299,51 +301,95 @@ func performEtcdStaticPodUpgrade(waiter apiclient.Waiter, pathMgr StaticPodPathM
+				return fmt.Errorf("failed to create etcd client for external etcd: %v", err)
+			}
+			oldEtcdClient = client
+			// Since etcd is managed externally, the new etcd client will be the same as the old client
```
I like that this structure is explicit because it makes interesting upgrade options possible for `externalEtcd -> staticPodEtcd` and vice versa.

stealthybox commented on cmd/kubeadm/app/util/etcd/etcd.go:
```diff
@@ -86,55 +83,86 @@ FlagLoop:
+		tlsInfo := transport.TLSInfo{
+			CertFile:      filepath.Join(certFile),
+			KeyFile:       filepath.Join(keyFile),
+			TrustedCAFile: filepath.Join(caFile),
```
`filepath.Join` looks like a noop here

detiber replied:
Yeah, I think this was a victim of copypasta, fixing.

stealthybox commented on cmd/kubeadm/app/phases/upgrade/staticpods_test.go:
```diff
@@ -208,33 +209,50 @@ func (spm *fakeStaticPodPathManager) BackupEtcdDir() string {
-	return etcdutil.PodManifestsHaveTLS(cluster.ManifestDir)
+func (c fakePodManifestEtcdClient) HasTLS() bool {
+	hasTLS, _ := etcdutil.PodManifestsHaveTLS(c.ManifestDir)
+	return hasTLS
```
for the test, we should add a TLS var to the struct and implement `NewFakePodManifestEtcdClient()` that pre-initializes it so that it behaves more like `etcdutil.GenericClient` (which does not handle podManifest mutations without recreation of the client)

detiber replied:
Nearing EOD here, but will try to see if I can find some time tonight to try and improve the tests around this.

stealthybox commented:
confirmed tests:
```
~/go/src/k8s.io/kubernetes/cmd/kubeadm detiber-stealthybox_upgrade
❯ bazel test //cmd/kubeadm/... >& /dev/null && echo tests passed
tests passed
```

stealthybox commented:
expecting this to pass:
```bash
# /vagrant/testTLSUpgrade.sh
set -ex

kubeadm reset \
  && kubeadm init --kubernetes-version 1.9.1 \
  && kubectl apply -f https://git.io/weave-kube-1.6

# user manually edits ConfigMap
kubectl -n kube-system edit cm kubeadm-config

sleep 300 # wait for CNI and Node Ready
/vagrant/bin/waitEtcd_kubeadm upgrade apply --yes 1.9.6  # non-TLS to TLS
sleep 120 # wait for cluster components to stabilize
/vagrant/bin/waitEtcd_kubeadm upgrade apply --yes 1.10.0 # TLS to TLS
sleep 120 # wait for cluster components to stabilize
/vagrant/bin/waitEtcd_kubeadm upgrade apply --yes 1.10.1 # TLS to TLS
sleep 120 # wait for cluster components to stabilize
/vagrant/bin/waitEtcd_kubeadm upgrade apply --yes 1.9.1  # TLS to TLS /w major version downgrade

echo
echo Test Passed
```

stealthybox commented:
The waiter worked:
```
[etcd] Wrote Static Pod manifest for a local etcd instance to "/etc/kubernetes/tmp/kubeadm-upgraded-manifests380775532/etcd.yaml"
[certificates] Generated etcd/ca certificate and key.
[certificates] Generated etcd/server certificate and key.
[certificates] etcd/server serving cert is signed for DNS names [localhost] and IPs [127.0.0.1]
[certificates] Generated etcd/peer certificate and key.
[certificates] etcd/peer serving cert is signed for DNS names [vagrant] and IPs [10.0.2.15]
[certificates] Generated etcd/healthcheck-client certificate and key.
[upgrade/staticpods] Moved new manifest to "/etc/kubernetes/manifests/etcd.yaml" and backed up old manifest to "/etc/kubernetes/tmp/kubeadm-backup-manifests456469723/etcd.yaml"
[upgrade/staticpods] Not waiting for pod-hash change for component "etcd"
[upgrade/etcd] waiting for etcd pod to become available after TLS upgrade
[util/etcd] Waiting 30s for initial delay
[util/etcd] Attempting to get etcd status 1/10
[util/etcd] Attempt failed with error: dial tcp [::1]:2379: connect: connection refused
[util/etcd] Waiting 15s until next retry
[util/etcd] Attempting to get etcd status 2/10
[util/etcd] Attempt failed with error: dial tcp [::1]:2379: connect: connection refused
[util/etcd] Waiting 15s until next retry
[util/etcd] Attempting to get etcd status 3/10
[util/etcd] Attempt failed with error: dial tcp [::1]:2379: connect: connection refused
[util/etcd] Waiting 15s until next retry
[util/etcd] Attempting to get etcd status 4/10
[upgrade/staticpods] Writing new Static Pod manifests to "/etc/kubernetes/tmp/kubeadm-upgraded-manifests380775532"
[controlplane] Wrote Static Pod manifest for component kube-apiserver to "/etc/kubernetes/tmp/kubeadm-upgraded-manifests380775532/kube-apiserver.yaml"
[controlplane] Wrote Static Pod manifest for component kube-controller-manager to "/etc/kubernetes/tmp/kubeadm-upgraded-manifests380775532/kube-controller-manager.yaml"
[controlplane] Wrote Static Pod manifest for component kube-scheduler to "/etc/kubernetes/tmp/kubeadm-upgraded-manifests380775532/kube-scheduler.yaml"
[certificates] Using the existing etcd/ca certificate and key.
[certificates] Generated apiserver-etcd-client certificate and key.
```

Pending other cases now.

detiber committed:
address some comments

stealthybox commented:
I ran the script twice and it passed all cases both times. 🎉 🌮

stealthybox commented on cmd/kubeadm/app/phases/upgrade/staticpods.go:
```diff
@@ -301,44 +301,48 @@ func performEtcdStaticPodUpgrade(waiter apiclient.Waiter, pathMgr StaticPodPathM
-		newEtcdClient = client
-	}
+	if isTLSUpgrade {
+		fmt.Println("[upgrade/etcd] waiting for etcd pod to become available after TLS upgrade")
```
If we diff this against master, I believe the L7 etcd check is always supposed to run regardless of TLS Upgrade.
The only difference would be we now have a waiter for extra safety.

fmt.Println("[upgrade/etcd] waiting for etcd to become healthy")

We could consider setting the initial delay only if isTLSUpgrade though since it's to avoid a race specific to not waiting for podHash

detiber replied:
Ah, indeed. I was wondering why I had originally left the conditional out :joy:

detiber replied:
Latest commit has removed the conditional.

stealthybox reviewed (changes requested):
I think we need to remove that conditional

detiber committed:
remove conditional

detiber committed:
address comment

stealthybox committed:
[kubeadm] Fix Etcd Rollback

Fix `rollbackEtcdData()` to return error=nil on success
`rollbackEtcdData()` used to always return an error making the rest of the
upgrade code completely unreachable.

Ignore errors from `rollbackOldManifests()` during the rollback since it
always returns an error.
Success of the rollback is gated with etcd L7 healthchecks.

stealthybox committed:
[kubeadm] Fine-tune delays for etcd waiters in upgrade

detiber committed:
Merge pull request #1 from stealthybox/fix_etcd_rollback

[kubeadm] Fix Etcd Rollback
